### ISSUE TITLE
SslStream using intermediate CA certificates

### ISSUE THREAD

**Issue by pepone** [dotnet/corefx#25581]
The SslStream class on Linux doesn't use the user personal certificate store to build the certificate chain

I have create a simple client server test to reproduce the issue https://github.com/pepone/netstreamtest

**Comment by karelz** [comment]
@Priya91 can you please take a look at this one? It seems to be difference between Windows and Linux.
cc @wfurt

**Comment by Priya91** [comment]
@pepone Reiterating to see if i understood you correctly, in this piece of code, 

```c#
// The following method is invoked by the RemoteCertificateValidationDelegate.
        public static bool ValidateServerCertificate(object sender,
                                                     X509Certificate certificate,
                                                     X509Chain chain,
                                                     SslPolicyErrors sslPolicyErrors)
        {
            Console.WriteLine("Chain length: {0}", chain.ChainElements.Count);
            var newChain = new X509Chain(false);
            newChain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
            newChain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
            // Use our Self signed CA certificate to build the newChain
            newChain.ChainPolicy.ExtraStore.Add(new X509Certificate2("cacert1.der"));
            newChain.Build(new X509Certificate2(certificate));

            Console.WriteLine("New Chain length: {0}", newChain.ChainElements.Count);
            if(sslPolicyErrors == SslPolicyErrors.None)
                return true;

            Console.WriteLine("Certificate error: {0}", sslPolicyErrors);

            // Do not allow this client to communicate with unauthenticated servers.
            return false;
        }
```

You care saying when you added the custom ```ExtraStore``` value of ```new X509Certificate2("cacert1.der")```, it doesn't use this while building the chain for ```new X509Certificate2(certificate)```? If that's the case, this is not a bug with SslStream, but X509Certificates.

cc @bartonjs

**Comment by bartonjs** [comment]
@pepone Does everything work as you expect if you import the intermediate certificate into the CurrentUser\CertificateAuthority store instead of CurrentUser\My on Linux?

**Comment by pepone** [comment]
@Priya91 I saying that the client should have receive the intermediate certificate but it only gets the server certificate. I updated the sample to allow running client and server on different machines, I run the server on a Windows machine and a separate macOS run the client in this case the client received chain is 2 indicating that the server send the intermediate certificate as this certificate is unknown for the client.

@bartonjs It doesn't make any difference, the client received chain lengh is still 1

```jose@cassiopeia:~/netstreamtest/client$ dotnet bin/Debug/netcoreapp2.0/client.dll 
Client connected.
Chain length: 1
New Chain length: 1
Certificate error: RemoteCertificateChainErrors
Exception: The remote certificate is invalid according to the validation procedure.```

**Comment by Priya91** [comment]
@pepone The bug is on Linux or MacOS? We have different codepaths for both. On Linux, when using Sslstream as server we do add the CA list from the ```LocalMachine``` and ```CurrentUser``` stores. What is the exact path where you have installed your intermediate certificate on your server Linux machine? Can you check if that location is same as values of ```StoreLocation.LocalMachine``` and ```StoreLocation.CurrentUser``` .NET APIs.

**Comment by pepone** [comment]
Hi @Priya91  I did a bit more testing and seems that the problem is on the Linux client not correctly building the keychain.

Windows Server | MacOS client the client keychain length is 2 
Linux Server | MacOS client the client keychain length is 2 
Linux Server | Linux client the client keychain length is 1

**Comment by Priya91** [comment]
I debugged this locally on linux, and it looks like the remote certificate chain obtained from openssl doesn't have the intermediate certificate. The code [here](https://github.com/dotnet/corefx/blob/master/src/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs#L77) gives the count of the chain stack as 1, and it only has the CA cert there, doesn't contain the intermediate cert.

Whereas on windows, the chain stack from the schannel contains 2 certs.

@bartonjs Can you help with this issue, do you know of any bugs with openssl where the cert chain is not received fully using ```SSL_get_peer_cert_chain```?

**Comment by bartonjs** [comment]
> The code here gives the count of the chain stack as 1, and it only has the CA cert there, doesn't contain the intermediate cert.

I assume you mean only has the leaf/EE/server cert.

> do you know of any bugs with openssl where the cert chain is not received fully using SSL_get_peer_cert_chain?

Nope, that should imply that the server just didn't send it.

Do I know of any cases where it won't send it?  Well, it gets built at https://github.com/dotnet/corefx/blob/a77d5aae4790689b86a9910ca981825b72daa9a8/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs#L146-L149.  Presumably the chain object at that state only has the one item in it.

**Comment by Priya91** [comment]
@pepone Confirms that the server sends the chain, with 

Windows Server | MacOS client the client keychain length is 2
Linux Server | MacOS client the client keychain length is 2
Linux Server | Linux client the client keychain length is 1

**Comment by Priya91** [comment]
@pepone Can you also have a Windows server and linux client, and confirm the server sends the entire chain, and the client fails to receive it..

**Comment by Priya91** [comment]
I did the following tests:

Windows Server | Linux client the client keychain length is 2
Linux Server | Linux client the client keychain length is 1

Also I've verified that the chain length is only 1 at the server side where the chain gets built, as linked above. It looks like the certificates are not correctly installed on linux to the openssl store. I opened an issue #26020 isolating the issue.

Closing this issue as the bug is tracked in the #26020 issue.
